### PR TITLE
#2325 with tweaks.

### PR DESCRIPTION
**Pull request by MistakeNot4892**
## Description of changes
Commit from #2325 with the requested changes. Quard seems to be gone from the Discord so unsure if he will implement them, figured I'd do it myself.

## Why and what will this PR improve
Character preview in prefs is larger and cleaner, and pAIs get their own preview when Global is selected.

## Authorship
@quardbreak for the main changes, I just tweaked a couple things.

## Changelog
:cl:
tweak: Character preview is a bit bigger and nicer.
/:cl:

### CODE REVIEW

File: code/modules/client/preference_setup/preference_setup.dm [hunk 1]
```diff
@@ -55,6 +55,8 @@ var/global/const/CHARACTER_PREFERENCE_INPUT_TITLE = "Character Preference"
 	sort_order = 10
 	category_item_type = /datum/category_item/player_setup_item/player_global
 
+/datum/category_group/player_setup_category/global_preferences/selected(var/datum/preferences/prefs)
+	prefs.update_pai_preview()
```
Comment (afterthought2) [inline code/modules/client/preference_setup/preference_setup.dm:59]: This call seems illegal (no args, so it still makes a blank mutable appearance and does something with it).

File: code/modules/client/preference_setup/preference_setup.dm [hunk 2]
```diff
@@ -115,9 +117,10 @@ var/global/const/CHARACTER_PREFERENCE_INPUT_TITLE = "Character Preference"
 		return 1
 
 	if(href_list["category"])
-		var/category = locate(href_list["category"])
-		if(category && (category in categories))
+		var/datum/category_group/player_setup_category/category = locate(href_list["category"])
+		if(istype(category) && (category in categories))
 			selected_category = category
+			category.selected(user.client.prefs)
```
Comment (afterthought2) [inline code/modules/client/preference_setup/preference_setup.dm:123]: I think `user.client.prefs` is actually wrong; these retain a ref to their prefs datum.

File: code/modules/client/preference_setup/preference_setup.dm [hunk 3]
```diff
@@ -248,7 +254,7 @@ var/global/const/CHARACTER_PREFERENCE_INPUT_TITLE = "Character Preference"
 	if(!pref_mob || !pref_mob.client)
 		return 1
 	if(. & TOPIC_UPDATE_PREVIEW)
-		pref_mob.client.prefs.update_preview_icon()
+		pref_mob.client.prefs.update_character_preview_map()
```
Comment (afterthought2) [inline code/modules/client/preference_setup/preference_setup.dm:257]: Again, should just be the instance's held prefs ref.

File: code/modules/mob/new_player/preferences_setup.dm [hunk 4]
```diff
@@ -98,12 +98,21 @@
...
-/datum/preferences/proc/update_preview_icon()
+/// Updates character mannequin without rebuilding the inventory or overlays of the mannequin.
+/datum/preferences/proc/refresh_character_preview()
+	var/mob/living/carbon/human/dummy/mannequin/mannequin = get_mannequin(client_ckey)
+	if(mannequin)
+		update_character_preview_mannequin(new /mutable_appearance(mannequin))
+
+/// Updates character preview map.
+/// If you don't need to update renders, use [/datum/preferences/proc/update_character_preview_background].
+/datum/preferences/proc/update_character_preview_map()
```
Comment (afterthought2) [inline code/modules/mob/new_player/preferences_setup.dm:109]: Why rename this? It's confusing; this proc doesn't actually do anything with the map control that any of the other split out ones don't do; the main thing it does is the expensive step of deleting and rebuilding the mannequin inventory.

The number of procs involved in this whole business feels needlessly confusing.

**Review by afterthought2 — COMMENTED** [review]
I can't actually follow what's happening in the code relative to when the pAI vs. the mob is shown to the user because of how it's set up: whichever call you make last will win, so it just depends on the ordering of the calls. As such, I have no idea if this is logically sound.

I think it would be much saner to have a single call (kind of like we do in atom code with `update_icon`) that you make to correctly display whatever should be shown based on prefs state, be that the pAI, the mannequin, nothing, whatever. It might be coupled with possibly one (1) auxiliary public function that's used to rebuild the preview mannequin (i.e. delete and rebuild the inventory) to alleviate performance concerns.

**Review by afterthought2 — CHANGES_REQUESTED** [review]
Actually I understated my objections in the previous comment. Looking over this more carefully and testing locally, the _only reason it actually works_ is the pAI update call in the html content getter for the global prefs tab. This should not even be there; that proc is not a refresh proc and should not be messing with client state (and defeats the purpose of the system). It also obfuscates that some of the other refresh calls are broken (see my comment), it obfuscates the fact that some calls from the global state menu will currently rebuild the mannequin, display the mannequin briefly (because of the way prefs/Topic and base Topic are set up, and it's visible to me even testing locally with no latency), and then replace that with the pAI appearance afterwards when the callstack arrives at that `contents` proc.

File: code/modules/client/preference_setup/global/03_pai.dm [hunk 5]
```diff
@@ -30,8 +27,7 @@
 	if(!candidate)
 		candidate = new()
 
-	if (!pai_preview)
-		update_pai_preview(user)
+	pref.update_pai_preview(candidate.chassis)
```
Comment (afterthought2) [inline code/modules/client/preference_setup/global/03_pai.dm:30]: This is what I was objecting to; this does not belong in this proc.

**Comment by MistakeNot4892** [comment]
Honestly I was adopting this because I thought it was alright, I'm not willing to revise it to the extent requested.

**Pull request closed without merging**